### DISCUSSION TITLE
Roslyn analyzer and cross-csproj checks.

### DISCUSSION THREAD

**Discussion by lsoft** [dotnet/roslyn#64273] [Q&A] [answered]
Hi!

I've created a Roslyn analyzer that adds `sealed` modifier if the class allows such action.

Due to I need to verfiy that my class are not parent of some different class, I registered `RegisterCompilationStartAction` and `RegisterSymbolAction` with `SymbolKind.NamedType` to collect  information about assembly types, and `RegisterCompilationEndAction` for analyzis and emitting diagnostics.

Everything works fine, except the following scenario:

I have a bug with 2 assemblies: assembly 1 with `public class Class1 {}` and assembly 2 with reference to assembly 1 and `class Class2 : Class1 {}`. I see the warning `AddSealed: Type name 'Class1' can be sealed` (**incorrect**) and the warning `AddSealed: Type name 'Class2' can be sealed` (**correct**).

So I need cross-assembly analysis somehow, any ideas are very welcome!
Thanks!

**Comment by sharwell** [comment, marked as answer]
Cross-assembly analysis is not currently supported by the analyzer driver. There are two options to handle a situation like this:

1. Create a custom command-line tool that uses Roslyn to analyze an entire solution at once instead of just looking at individual projects.
2. Update the analyzer to ignore types that are accessible outside the current assembly.

Most of our internal analyzers use the second option.

**Reply by lsoft** [reply]
@sharwell got it, thank you. Could you share: are there any plans to include cross-assembly analysis (whole solution analysis?) in the driver?

**Reply by sharwell** [reply]
I'm not aware of any such plans. @mavasani may have more details.

**Reply by mavasani** [reply]
There are no current plans for supporting solution-level/cross-project analysis that I am aware of. Tagging @mikadumont

**Reply by lsoft** [reply]
Ecosystem will hugely benefit from such feature! I'm sure there are a lot of ideas which can not be fully realized without such analysis.
